index.js

```diff
@@ -267,9 +267,25 @@ const execa = (command, args, options) => {
 		}, parsed.options.timeout);
 	}
 
+	const resolvable = (() => {
+		let extracted;
+		const promise = new Promise(resolve => {
+			extracted = resolve;
+		});
+		promise.resolve = extracted;
+		return promise;
+	})();
+
 	const processDone = new Promise(resolve => {
 		spawned.on('exit', (code, signal) => {
 			cleanup();
+
+			if (timedOut) {
+				resolvable.resolve([
+					{code, signal}, '', '', ''
+				]);
+			}
+
 			resolve({code, signal});
```

Comment:
I am wondering whether the following would be simpler (did not try, this is just an example):

```js
	const timeoutPromise = new Promise(resolve => {
       if (parsed.options.timeout > 0) {
		    timeoutId = setTimeout(() => {
			    timeoutId = undefined;
			    timedOut = true;
                resolve([{ code: '1', signal: parsed.options.killSignal }, '', '', '']);
		        spawned.kill(parsed.options.killSignal);
		    }, parsed.options.timeout);
        }
	});

    // No changes in the code there
	const processDone = new Promise(resolve => {
		spawned.on('exit', (code, signal) => {
			cleanup();
			resolve({code, signal});
		});

		spawned.on('error', error => {
			cleanup();
			resolve({error});
		});

		if (spawned.stdin) {
			spawned.stdin.on('error', error => {
				cleanup();
				resolve({error});
			});
		}
	});

	const handlePromise = () => pFinally(Promise.race([
        timeoutPromise,
        Promise.all([
		    processDone,
	   	    getStream(spawned, 'stdout', {encoding, buffer, maxBuffer}),
		    getStream(spawned, 'stderr', {encoding, buffer, maxBuffer}),
		    getStream(spawned, 'all', {encoding, buffer, maxBuffer: maxBuffer * 2})
	    ]).then(results => { // eslint-disable-line promise/prefer-await-to-then
```

Reply:
Thinking about it, it's probably a bad idea as the main promise will resolve before the child process has exited.

Reply:
Yes, this works and look cleaner to me. One thing tho, I'm not a fan of never resolving promises, but I might be overly protective.

Reply:
The bigger issue with my suggestion was that the main promise would be resolved before the child process has exited, so you can discard it :)

Reply:
I can make a mix of both world with:

```js
// TODO: Use native "finally" syntax when targeting Node.js 10
const handlePromise = () => {
	const timedOutPromise = new Promise(resolve => {
		if (parsed.options.timeout > 0) {
			timeoutId = setTimeout(() => {
				timeoutId = undefined;
				timedOut = true;
				const {killSignal} = parsed.options;
				resolve([
					{code: 1, signal: killSignal},
					'',
					'',
					''
				]);
				spawned.kill(killSignal);
			}, parsed.options.timeout);
		}
	});

	return pFinally(Promise.race([
		timedOutPromise,
		Promise.all([
			processDone,
			getStream(spawned, 'stdout', {encoding, buffer, maxBuffer}),
			getStream(spawned, 'stderr', {encoding, buffer, maxBuffer}),
			getStream(spawned, 'all', {encoding, buffer, maxBuffer: maxBuffer * 2})
		])
	]).then(results => { // eslint-disable-line promise/prefer-await-to-then
    // ...
```

Reply:
No the problem there is that `resolve()` gets called before the `exit` event gets emitted. I.e. `execa()` will resolve before the child process exited.